File: compare_python_matlab/compute_quality.m
load('generated_data') % loads X, codebook0, codebook1, codebook2, codebook3, IDX1, SUMD

[num_cent num_feat] = size( codebook0 );
all_codebooks = zeros(num_cent,num_feat,4);
all_codebooks(:,:,1) = codebook0;
all_codebooks(:,:,2) = codebook1;
all_codebooks(:,:,3) = codebook2;
all_codebooks(:,:,4) = codebook3;

addpath('../braun_geurten_egelhaaf/Eval');

instability = pbc_calcCentroidInstability( all_codebooks );

[qualities] = pbc_calcClusterQuality(codebook0, IDX1, SUMD);
quality = qualities(:,4);

disp(sprintf("instability: %.5f",instability));
disp(sprintf("quality: [%.5f, %.5f]",quality(1), quality(2)));

File: compare_python_matlab/compute_quality.py
import numpy as np
import scipy.io
from scipy.cluster.vq import vq, kmeans, whiten
import sys, os

if 1:
    # even if postcluster is not installed, put it in path
    mydir = os.path.split( os.path.abspath(__file__))[0]
    postcluster_dir = os.path.abspath(os.path.join(mydir,'..'))
    sys.path.insert(0, postcluster_dir )

import postcluster

mdata = scipy.io.loadmat('generated_data.mat',struct_as_record=True)
X = mdata['X']
codebook0 = mdata['codebook0']
codebook1 = mdata['codebook1']
codebook2 = mdata['codebook2']
codebook3 = mdata['codebook3']

all_codebooks = np.array( [codebook0, codebook1, codebook2, codebook3] )
instability_results = postcluster.match_centroids( all_codebooks )
instability = instability_results['mean_instability']

quality = postcluster.get_quality( codebook0, X )

print "instability: %.5f"%instability
print "quality: [%.5f, %.5f]"%(quality[0], quality[1])
print
